File: haskell/src/Solutions/Euler023.hs
module Solutions.Euler023 where
import qualified Data.Set as Set

listPlus l1 l2 = (+) <$> l1 <*> l2

bound = 28123 :: Int
divisors :: Int -> [Int]
divisors n = filter ((==0) . mod n) [1..n - 1]
abundant :: [Int]
abundant = filter (\n -> sum (divisors n) > n) [1..bound]

euler023 :: IO String
euler023 = return $ show $ sum $ Set.toList
         $ Set.fromList [1..bound] Set.\\ Set.fromList (listPlus abundant abundant)

File: rust/src/solutions/euler023.rs
use crate::utils::numeric::divisors::divisors;

pub fn euler023() -> String {
    let bound: usize = 28124;
    let abdunant: Vec<usize> = (1..bound)
        .filter(|&n| divisors(n).into_iter().sum::<usize>() > n << 1)
        .collect();
    let mut possible: Vec<bool> = vec![false; bound];
    for i in &abdunant {
        for j in &abdunant {
            let s = i + j;
            if s < bound {
                possible[s] = true;
            }
        }
    }
    possible
        .into_iter()
        .zip(0..)
        .filter_map(|(t, i)| if t { None } else { Some(i) })
        .sum::<usize>()
        .to_string()
}
